modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,403 @@
+	 * It also checks if max number of attempts have been attempted for the job.
+	 * For any such condition, it will return false, else true.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool Flag to tell whether this job should run or not.
+	 */
+	public function should_run() {
+		// If job doesn't exist or completed, return false.
+		if ( ! $this->exists() || $this->is_completed() ) {
+			return false;
+		}
+
+		if ( $this->is_running() ) {
+			return false;
+		}
+
+		/**
+		 * Number of attempts to try to run a job.
+		 * Repeated attempts may be required to run a failed job. Default 3 attempts.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param int $attempts Number of attempts allowed for a job to run.
+		 */
+		$max_attempts = apply_filters( 'perflab_job_max_attempts_allowed', 3 );
+
+		// If number of attempts have been exhausted, return false.
+		if ( $this->get_attempts() >= absint( $max_attempts ) ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Sets the status of the current job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $status Status to set for current job.
+	 * @return bool Returns true if status is updated, false otherwise.
+	 */
+	public function set_status( $status ) {
+		$valid_statuses = array(
+			self::JOB_STATUS_COMPLETE,
+			self::JOB_STATUS_FAILED,
+			self::JOB_STATUS_QUEUED,
+			self::JOB_STATUS_RUNNING,
+			self::JOB_STATUS_PARTIAL,
+		);
+
+		if ( in_array( $status, $valid_statuses, true ) ) {
+			update_term_meta( $this->job_id, self::JOB_STATUS_META_KEY, $status );
+
+			// If job is complete, set the timestamp at which it was completed.
+			if ( self::JOB_STATUS_COMPLETE === $status ) {
+				update_term_meta( $this->job_id, 'job_completed_at', time() );
+			}
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Records the error for the current process run.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param WP_Error $error Error instance.
+	 *
+	 * @todo Implement the error recording activity.
+	 *
+	 * @return void
+	 */
+	public function set_error( WP_Error $error ) {
+		$job_failure_data = $error->get_error_data( 'perflab_job_failure' );
+
+		if ( ! empty( $job_failure_data ) ) {
+			$this->set_status( self::JOB_STATUS_FAILED );
+
+			update_term_meta( $this->job_id, self::JOB_ERRORS_META_KEY, $job_failure_data );
+			update_term_meta( $this->job_id, self::JOB_ATTEMPTS_META_KEY, ( $this->get_attempts() + 1 ) );
+		}
+	}
+
+	/**
+	 * Retrieves the job status from its meta information.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Job status.
+	 */
+	public function get_status() {
+		return (string) get_term_meta( $this->job_id, self::JOB_STATUS_META_KEY, true );
+	}
+
+	/**
+	 * Retrieves the job data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array|null Job data.
+	 */
+	public function get_data() {
+		// If we have cached data, return it.
+		if ( ! is_null( $this->data ) ) {
+			return $this->data;
+		}
+
+		$this->data = get_term_meta( $this->job_id, self::JOB_DATA_META_KEY, true );
```

Comment:
Why do we need the logic here to look up the term data if it isn't set? Shouldn't it be set already by the constructor?

If there is a good reason that we should have this logic here, why don't we do the same for the below `get_name()` method?

Reply:
This was to avoid the extra function call in the call stack, but I don't think this has much of the impact as `get_term_meta` itself has caching mechanism as you mentioned above.
Removing this condition.